Sign ProRegTx collaterals with a string message instead of payload hash, split `protx register` into `prepare`/`submit`

codablock opened the pull request:
Required for HW wallet support. See https://github.com/dashpay/dips/pull/28

codablock commented:
Hmm I'm currently not happy with including the magic string "DarkCoin Signed Message:\n" in the message. Should we maybe change `CMessageSigner` to allow signing without the magic str or with a specific magic str for the use case?

Bertrand256 commented:
@codablock, I missed what @UdjinM6 noticed: [protx_register](https://github.com/dashpay/dash/blob/96e6f1c1c48977b5d5535a2eea0a7d5465373e98/src/rpc/rpcevo.cpp#L214-L231
) function doesn't accept parameter for the `payloadSig` field of ProRegTx transaction.

[Here](https://github.com/dashpay/dash/blob/96e6f1c1c48977b5d5535a2eea0a7d5465373e98/src/rpc/rpcevo.cpp#L342-L345) is the assumption, that the collateral key is controlled by the the local node's wallet, but we also should use the `payloadSig` if provided as an argument.

codablock commented:
@Bertrand256 @UdjinM6 My expectation was that the external tool which does the signature would also have to fully implement serialization of the payload and TX. Otherwise it wouldn't be able to calculate the payload hash to actually create the message to sign, as it also includes the payload hash. If the external tool creates the whole transaction, then it can also pass the fully signed TX to the network.

UdjinM6 commented:
> My expectation was that the external tool which does the signature would also have to fully implement serialization of the payload and TX.

Yep, this was smth I assumed as well but _it seems_ that HW wallets won't be able to sign this "weird" raw tx because there is no way to how to deserialize/hash/sign txes with data they don't recognize. For trezor the flow is https://wiki.trezor.io/Developers_guide:Message_Workflows#SignTx. Note that there is TXEXTRADATA for zcash but it's not working the way we need, it's a per-input data as far as I was able to understand. This means that the only way to sign such tx (without modifying HW wallet firmware) is Dash Core (or whatever wallet/tool will be able to sign special txes). So the workflow I see atm would be smth like this:
1. create non-signed tx with empty payloadSig via `proregtx` (`proregtx create`?) in Dash Core, outputs rawtx + payload "text" to sign;
2. sign the payload "text" via `SignMessage` in some external tool (could be DMT or even official wallet app/page);
3. sign and submit proregtx by providing rawtx from step 1 and sig from step 2 (`proregtx submit`?).

The process is far from being user friendly unfortunately but I don't see how to simplify it atm.

Bertrand256 commented:
@UdjinM6 Why do we need step 1? Is that non-signed tx to be used along with the payload data to create a signature in step 2 and thus act as a protection against replay or smth else?

codablock commented:
Added 2 commits:
1. Rename `fund_register` to `register_fund` (better grouping of commands)
2. Implemented `register_prepare` and `register_submit`. `register_prepare` expects the same args as the other 2 register commands but returns an unsigned ProRegTx and the message to sign. You can then use `signmessage` (or a HW wallet) with the collateral address and the message to sign that message. `register_submit` is then used to finalize and submit the TX.

Also rebased on develop.

UdjinM6 commented on src/rpc/rpcevo.cpp:
```diff
@@ -762,6 +840,8 @@ UniValue protx(const JSONRPCRequest& request)
                 "  register          - Create and send ProTx to network\n"
                 "  register_fund     - Fund, create and send ProTx to network\n"
+                "  register_prepare  - Create an unsigned ProTx\n"
+                "  register_submit   - Signs and submits a ProTx\n"
```
nit: `Sign and submit` (no `s`) to keep it consistent

UdjinM6 commented:
@Bertrand256 we need step 1 to construct the payload data
